### PR TITLE
GitHub Action workflow to detect invalid domain input

### PR DESCRIPTION
**Pull request by xm1k3**
```
Invalid domains found: achmea.com.au
eurekosigorta.com.tr
gainsboroughhardware.com.au
fshlocking.com.au
allegion.co.uk
brio.com.au
```
https://github.com/projectdiscovery/public-bugbounty-programs/actions/runs/4118386653

### CODE REVIEW

File: .github/workflows/validate-domains.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v3
+    - name: Validate domains
+      run: |
+        invalid_domains=()
+        tmpfile=$(mktemp)
+        jq -r '.programs[].domains[]' chaos-bugbounty-list.json > "$tmpfile"
+        while IFS= read -r tld; do
+            if [[ "$tld" != $(echo "$tld" | awk -F . '{print $(NF-1)"."$NF}') ]]; then
```
Comment (Mzack9999) [inline .github/workflows/validate-domains.yml:21]: Check for '.' presence might not suffice. A tiny cli utility that validates the domains would be a better option, as we need to cover all the cases listed in the issue (e.g. URL, wildcard, etc.). Among the many possible approaches, for example, we can use [IsDNSName](https://github.com/asaskevich/govalidator).
As we are processing json lines, it can also validate the JSON 'URL' field for a valid URL.

**Review by Mzack9999 — CHANGES_REQUESTED** [review]

File: cmd/validate-domains/main.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+func main() {
+	flag.Parse()
+
+	rawJSON, err := ioutil.ReadFile(*bbListFile)
+	if err != nil {
+		log.Fatalf("Failed to read initial JSON file: %v", err)
+	}
+
+	var ps programs
```
Comment (Mzack9999) [inline cmd/validate-domains/main.go:35]: I'd suggest using `https://github.com/tidwall/gjson`, the whole unmarshaling could be simplified to:

```go
gdata := gjson.ParseBytes(rawJSON)
gdata.Get("programs.#.domains|@flatten").ForEach(func(key, value gjson.Result) bool {
        // domain validation logic here
        domain := value.String()
        ...
	return true
})
```

File: cmd/validate-domains/main.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+			if strings.Count(domain, ".") > 1 {
+				invalidDomains = append(invalidDomains, domain)
+			}
+		}
+	}
+
+	output := strings.Join(invalidDomains, "\n")
+
+	err = ioutil.WriteFile("invalid_domains.txt", []byte(output), 0644)
```
Comment (Mzack9999) [inline cmd/validate-domains/main.go:56]: `os.WriteFile`

**Comment by xm1k3** [comment]
@Mzack9999 can it be a good idea to have main programs under `cmd/` and all functions directly under `/pkg/core` like i did?

**Comment by Mzack9999** [comment]
@xm1k3 By convention, in other repositories, the tools are under `/cmd/xxx`, so we can even move the existing `bbp-scope` into `/cmd/bbp-scope` (as it seems not used anywhere).
Regarding package organization, I'd suggest splitting them into groups into subfolders of `/pkg`. In this case, as dns it's being handled, it could be, for example, `/pkg/dns`

File: pkg/core/core.go [hunk 4]
```diff
@@ -259,20 +251,20 @@ func readChaosBountyPrograms() (map[string]chaosProgram, error) {
...
+	chaosMap := make(map[string]ChaosProgram)
 	for _, value := range list.Programs {
 		chaosMap[value.Name] = value
 	}
 	log.Printf("[INFO] Read %d programs from chaos list\n", len(chaosMap))
 	return chaosMap, nil
 }
 
-func extractHostname(item string) string {
+func ExtractHostname(item string) string {
```
Comment (Mzack9999) [inline pkg/core/core.go:267]: This function does more than just validating if `item` is a valid domain (eg. converts to lowercase and strip prefixes). I'd suggest to move the anonymous function `validate` into a package exported function (eg. `ValidateFQDN`), since it contains the logic that we need.

File: cmd/validate-domains/main.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	var ps core.ChaosList
+	err = json.Unmarshal(rawJSON, &ps)
+	if err != nil {
+		log.Fatalf("Failed to parse initial JSON file: %v", err)
+	}
+
+	var invalidDomains []string
+	gdata := gjson.ParseBytes(rawJSON)
+	gdata.Get("programs.#.domains|@flatten").ForEach(func(key, value gjson.Result) bool {
+		domain := value.String()
```
Comment (Mzack9999) [inline cmd/validate-domains/main.go:33]: The empty value would be considered valid with this check as `core.ExtractHostname("")` returns `""`. Consider implementing tabular-tests to ensure that most basic edge cases are covered by the logic. An incomplete list can be:
- [ ] Empty value (`""`)
- [ ] Spaces (`" "`)
- [ ] contains wildcard ("*")
- [ ] URL (`"http://`, `https://`)
- [ ] non primary/secondary dns name (`xxx.projectdiscovery.io`)
- [ ] ...

**Review by Mzack9999 — CHANGES_REQUESTED** [review]

**Review by Mzack9999 — APPROVED** [review]
